src/raw/mod.rs

```diff
@@ -279,7 +267,9 @@ impl<T> Bucket<T> {
             // won't overflow because index must be less than length
             (index + 1) as *mut T
         } else {
-            base.as_ptr().add(index)
+            // the pointer arithmetic below might cross allocation bounds
+            // because RawTable::iter() could call this function with empty table and index=0
+            base.as_ptr().wrapping_add(!index) as usize as *mut T
```

Comment:
Why aren't you just using `base.as_ptr().sub(index).sub(1)`? This should not overflow.

Reply:
`base.as_ptr().sub(index).sub(1)` crosses allocation boundaries when called with empty table which is UB. I also tried doing `base.as_ptr().sub(index)` and returning `self.ptr.as_ptr().sub(1)` but ended up being a lot slower (compared to this PR)
```
 name                         before ns/iter  after ns/iter  diff ns/iter   diff %  speedup 
 clone_from_large             1,530           1,490                   -40   -2.61%   x 1.03 
 clone_from_small             22              19                       -3  -13.64%   x 1.16 
 clone_large                  1,652           1,689                    37    2.24%   x 0.98 
 clone_small                  40              42                        2    5.00%   x 0.95 
 insert_ahash_highbits        9,582           9,140                  -442   -4.61%   x 1.05 
 insert_ahash_random          10,099          9,383                  -716   -7.09%   x 1.08 
 insert_ahash_serial          11,803          10,261               -1,542  -13.06%   x 1.15 
 insert_erase_ahash_highbits  56,871          21,035              -35,836  -63.01%   x 2.70 
 insert_erase_ahash_random    43,442          21,524              -21,918  -50.45%   x 2.02 
 insert_erase_ahash_serial    38,006          33,962               -4,044  -10.64%   x 1.12 
 insert_erase_std_highbits    85,009          50,708              -34,301  -40.35%   x 1.68 
 insert_erase_std_random      83,605          51,105              -32,500  -38.87%   x 1.64 
 insert_erase_std_serial      80,513          42,439              -38,074  -47.29%   x 1.90 
 insert_std_highbits          23,145          22,547                 -598   -2.58%   x 1.03 
 insert_std_random            23,268          22,267               -1,001   -4.30%   x 1.04 
 insert_std_serial            20,553          18,894               -1,659   -8.07%   x 1.09 
 iter_ahash_highbits          2,124           1,498                  -626  -29.47%   x 1.42 
 iter_ahash_random            2,126           1,361                  -765  -35.98%   x 1.56 
 iter_ahash_serial            2,130           1,429                  -701  -32.91%   x 1.49 
 iter_std_highbits            2,130           1,437                  -693  -32.54%   x 1.48 
 iter_std_random              2,125           1,354                  -771  -36.28%   x 1.57 
 iter_std_serial              2,123           1,488                  -635  -29.91%   x 1.43 
 lookup_ahash_highbits        5,512           4,683                  -829  -15.04%   x 1.18 
 lookup_ahash_random          5,503           4,610                  -893  -16.23%   x 1.19 
 lookup_ahash_serial          6,041           4,307                -1,734  -28.70%   x 1.40 
 lookup_fail_ahash_highbits   4,609           4,011                  -598  -12.97%   x 1.15 
 lookup_fail_ahash_random     4,823           4,488                  -335   -6.95%   x 1.07 
 lookup_fail_ahash_serial     6,632           7,706                 1,074   16.19%   x 0.86 
 lookup_fail_std_highbits     18,709          18,488                 -221   -1.18%   x 1.01 
 lookup_fail_std_random       18,695          18,422                 -273   -1.46%   x 1.01 
 lookup_fail_std_serial       16,261          15,339                 -922   -5.67%   x 1.06 
 lookup_std_highbits          19,321          18,104               -1,217   -6.30%   x 1.07 
 lookup_std_random            19,664          18,272               -1,392   -7.08%   x 1.08 
 lookup_std_serial            16,507          15,220               -1,287   -7.80%   x 1.08
```

Reply:
I tested this locally and the effect was much smaller. Did you forget to remove the `- 1` in `to_base_index`?

Reply:
I was using 2 weeks old nightly version, switching to stable or updated nightly fixed the results. Updated bench in first post.